src/v/archival/ntp_archiver_service.cc

```diff
@@ -1242,8 +1246,29 @@ ss::future<ntp_archiver::upload_group_result> ntp_archiver::wait_uploads(
             }
         }
 
+        // Only uploads of the latest segments update the next delta offset.
+        if (segment_kind == segment_upload_kind::non_compacted) {
+            auto ot_state = _parent.get_offset_translator_state();
+            auto segment_offset_next = model::next_offset(
+              upload.meta->committed_offset);
+            auto segment_delta_next = model::offset_delta(
+              segment_offset_next()
+              - ot_state->from_log_offset(segment_offset_next)());
+            if (delta_offset_next.has_value()) {
+                delta_offset_next = std::max(
+                  *delta_offset_next, segment_delta_next);
+            } else {
+                delta_offset_next = segment_delta_next;
+            }
+        }
         mdiff.push_back(*upload.meta);
     }
+    if (
+      delta_offset_next.has_value()
+      && *delta_offset_next
+           <= _parent.archival_meta_stm()->manifest().delta_offset_next()) {
+        delta_offset_next = std::nullopt;
+    }
```

Comment:
Hrm, I'm actually not too keen on this. While I agree we should add safeguards to the STM, IMO it shouldn't be the arbiter of whether to replicate or not. It can change the way it applies (and we do that by ignoring in `advance_delta_offset_next()`), but policy decisions about how to replicate feel like they belong in the upload loop.